### PR TITLE
feat(maintenance): drop support for Node.js 16.x

### LINKED ISSUE #2223: Maintenance: drop support for Node.js 16
### Summary

> [!important]
> Starting from July 1, 2024 Powertools for AWS Lambda (TypeScript) will end support for Node.js 16.

This action follows [the date when AWS Lambda will mark the `nodejs16.x` as deprecated](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html#runtimes-supported) and precedes the date after which you will no longer be able to create new functions using the runtime.

We recommend that you upgrade your existing Node.js 16 functions to Node.js 20 before July 1st, 2024.

End of support does not impact existing functions and functions using versions of Powertools for AWS Lambda (TypeScript) released before the date, however starting from this date we will not guarantee compatibility with Node.js 16.

You can use the following AWS CLI command to list all functions in a specific region using Node.js 16, including published function versions:

> [!Note]
> Make sure to replace the value of the `--region` flag and run this command in each region you have Lambda functions workloads

```sh
aws lambda list-functions --function-version ALL --region <Region> --output text --query "Functions[?Runtime=='nodejs16.x'].FunctionArn"
```

### Why is this needed?

Starting from July 15, 2024 we will no longer be able to deploy new Lambda functions using the `nodejs16.x` runtime, this means we will no longer be able to run integration tests for new code changes.

With this in mind, and considering our iteration cycles last two weeks, we plan on dropping support for the runtime just two weeks before the date.

### Which area does this relate to?

Governance

### Solution

_No response_

### Acknowledgment

- [X] This request meets [Powertools for AWS Lambda (TypeScript) Tenets](https://docs.powertools.aws.dev/lambda/typescript/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Python](https://github.com/aws-powertools/powertools-lambda-python/), [Java](https://github.com/aws-powertools/powertools-lambda-java/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### Future readers

Please react with 👍 and your use case to help us understand customer demand.

### PR DESCRIPTION
**Pull request by dreamorosi**
### Changes

> Please provide a summary of what's being changed

This PR drops support for Node.js 16 for Powertools for AWS Lambda (TypeScript). As discussed in the linked issue, Node.js 16 has reached EOL and its corresponding AWS Lambda managed runtime has been marked as deprecated.

Below a list of changes included in the PR:
- Remove Node.js 16 from test matrixes, test harnesses, scripts, etc.
- Remove deprecated version from docs, comments, issue/PR templates
- Removed deprecation banner from docs, replaced with v1 EOL
- Raised baseline for Layer canary stack to Node.js 18
- Removed `nodejs16x` from compatible runtimes in Lambda layer
- Replaced usage of `Object(someObject).hasOwnProperty(someProperty)` with `Object.hasOwn(someObject, someProperty)` ([MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn))
- Removed polyfill for Error.cause ([MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause)) and 
 - updated `ParseError` in Parser utility to use it, as well as every envelope
 - updated Logger log formatter to assume the option can exist
- Removed conditional behavior around `fetch` in e2e tests since now it's present in all supported runtimes
 
> Please add the issue number below, if no issue is present the PR might get blocked and not be reviewed

**Issue number:** closes #2223

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: packages/idempotency/tests/helpers/populateEnvironmentVariables.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // Reserved variables
 process.env._X_AMZN_TRACE_ID = '1-abcdef12-3456abcdef123456abcdef12';
 process.env.AWS_LAMBDA_FUNCTION_NAME = 'my-lambda-function';
-process.env.AWS_EXECUTION_ENV = 'nodejs16.x';
+process.env.AWS_EXECUTION_ENV = 'nodejs20.x';
```
Comment (am29d) [inline packages/idempotency/tests/helpers/populateEnvironmentVariables.ts:4]: should we put the default to lowest nodejst version as in other files?

File: packages/jmespath/src/errors.ts [hunk 2]
```diff
@@ -108,13 +108,6 @@ class ParseError extends JMESPathError {
...
  */
 class IncompleteExpressionError extends ParseError {
-  /**
-   * Expression that was being parsed when the error occurred.
-   *
-   * Can be set by whatever catches the error.
-   */
-  public expression?: string;
-
   public constructor(options: {
```
Comment (am29d) [inline packages/jmespath/src/errors.ts:111]: Is this related to the PR?

File: packages/testing/tests/helpers/populateEnvironmentVariables.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
 // Reserved variables
 process.env._X_AMZN_TRACE_ID = '1-abcdef12-3456abcdef123456abcdef12';
 process.env.AWS_LAMBDA_FUNCTION_NAME = 'my-lambda-function';
-process.env.AWS_EXECUTION_ENV = 'nodejs16.x';
+process.env.AWS_EXECUTION_ENV = 'nodejs20.x';
```
Comment (am29d) [inline packages/testing/tests/helpers/populateEnvironmentVariables.ts:4]: node18 as default?

**Review by am29d — CHANGES_REQUESTED** [review]
Great work on keeping the track on all the places where node16 had a special treatment! 

Few notes on default node version in various places, so we can settle on either 18 or 20.

File: packages/jmespath/src/errors.ts (hunk 2 above)
Comment (dreamorosi) [reply packages/jmespath/src/errors.ts:111]: Changing the `target` in `tsconfig.json` to Node.js 18 caused this issue to pop up.

This error extends another base error that has the same `expression` property, which was redeclared here, thus causing the error.

I should've added a comment in the PR to explain this, my bad.

File: packages/testing/tests/helpers/populateEnvironmentVariables.ts (hunk 3 above)
Comment (dreamorosi) [reply packages/testing/tests/helpers/populateEnvironmentVariables.ts:4]: This is just an environment variable for the unit tests, I decided to jump to 20 since it's the same we run on the dev environment when we run the tests.

File: packages/idempotency/tests/helpers/populateEnvironmentVariables.ts (hunk 1 above)
Comment (dreamorosi) [reply packages/idempotency/tests/helpers/populateEnvironmentVariables.ts:4]: Let's discuss it in [this thread](https://github.com/aws-powertools/powertools-lambda-typescript/pull/2717#discussion_r1662133826) and settle both.

File: packages/testing/tests/helpers/populateEnvironmentVariables.ts (hunk 3 above)
Comment (dreamorosi) [reply packages/testing/tests/helpers/populateEnvironmentVariables.ts:4]: This environment variable is used only to do two things:
- Know whether we need to append the UA middleware to a client ([here](https://github.com/arnabrahman/aws-lambda-powertools-typescript/blob/8eeb735c8333a5adb649defa47aa5759fbec24ce/packages/commons/src/awsSdkUtils.ts#L4))
- Know whether we need to disable the Tracer ([here](https://github.com/arnabrahman/aws-lambda-powertools-typescript/blob/8eeb735c8333a5adb649defa47aa5759fbec24ce/packages/tracer/src/config/EnvironmentVariablesService.ts#L9))

The variable is used as a proxy for detecting whether we are running within Lambda or elsewhere (i.e. local/CI). In both cases Powertools only cares to know if the variable is present or not, since it's set automatically by Lambda.

For this reason, since we are running Node.js 20x in our dev environment I set it to this version; also so that for when we drop Node.js 18 we have two files less to touch.

However, I can also see an argument for setting it to Node.js 18 for consistency - if we agree on this I'm okay to revert these two files.

Comment (am29d) [reply packages/testing/tests/helpers/populateEnvironmentVariables.ts:4]: Thanks for the context, this clears things up! Because there is no special case for node18, like we had for node16, I suggest we can move to node20 as a default. What do you think?

Comment (dreamorosi) [reply packages/testing/tests/helpers/populateEnvironmentVariables.ts:4]: Our development environment is already on Node.js 20, which other default do you mean?

Comment (dreamorosi) [reply packages/testing/tests/helpers/populateEnvironmentVariables.ts:4]: Talked about this offline & clarified that we are already using Node.js 20 for dev. Resolving.

Comment (am29d) [reply packages/testing/tests/helpers/populateEnvironmentVariables.ts:4]: dev, layer and target, mostly for consistency. So we keep layer and target at 18 and dev at 20 as it is now.

**Review by am29d — APPROVED** [review]
